feat(sdn-controller): use host plugin

fbeauchamp opened the pull request:
### Description

to enable the use of plugin, add this in a config file 

[plugins.sdn-controller]
useDirectChannel = false 

Includes this fix : https://github.com/vatesfr/xen-orchestra/pull/9000

plugin is here : https://github.com/xcp-ng/xcp-ng-xapi-plugins/blob/dml-traffic-rules-fixing/SOURCES/etc/xapi.d/plugins/sdncontroller.py

method can be called by xo-cli : sdnController.addRule and sdnController.deleteRule 

and 2 new methods : sdnController.addNetworkRule ans sdnController.deleteNetworkRule that take the same parameter as addRule and deleteRule but with a networkId instead of a vifId . **This 2 methods will fail on the direct (legacy) approach. They will only work in coordination with the plugin**


### Checklist

- Commit
  - Title follows [commit conventions](https://bit.ly/commit-conventions)
  - Reference the relevant issue (`Fixes #007`, `See xoa-support#42`, `See https://...`)
  - If bug fix, add `Introduced by`
- Changelog
  - If visible by XOA users, add changelog entry
  - Update "Packages to release" in `CHANGELOG.unreleased.md`
- PR
  - If UI changes, add screenshots
  - If not finished or not tested, open as _Draft_

### Review process

> This 2-passes review process aims to:
>
> - develop skills of junior reviewers
> - limit the workload for senior reviewers
> - limit the number of unnecessary changes by the _author_

1. The _author_ creates a PR.
2. Review process:
   1. The _author_ assigns the _junior reviewer_.
   2. The _junior reviewer_ conducts their review:
      - Resolves their comments if they are addressed.
      - Adds comments if necessary or approves the PR.
   3. The _junior reviewer_ assigns the _senior reviewer_.
   4. The _senior reviewer_ conducts their review:
      - If there are no unresolved comments on the PR → merge.
      - Otherwise, we continue with **3.**
3. The _author_ responds to comments and/or makes corrections, and we go back to **2.**

Notes:

1. The _author_ can request a review at any time, even if the PR is still a _Draft_.
2. In theory, there should not be more than one reviewer at a time.
3. The _author_ should not make any changes:
   - When a reviewer is assigned.
   - Between the _junior_ and _senior_ reviews.

fbeauchamp committed:
fix(sdn-controller): don't fail at failing

fbeauchamp committed:
feat(sdn-controller): use xapi plugin to manage open switch

try to instantiate the plugin and fall back to direct query
if it fails
The plugin allow to define and target multiple bridges in parallel

fbeauchamp committed:
feat(sdn-controller): implement network wide rules

this introduce 2 new API method sdnController.addNetworkRule
and sdnController.deleteNetworkRule
This method will only work with the plugin, and can't work with
the legacy ( direct channel) approach

b-Nollet commented on packages/xo-server-sdn-controller/src/openflow-controller.js:
```diff
@@ -0,0 +1,16 @@
+    await controller.check(host)
+    return controller
+  } catch (error) {
+    if (error.code === 'XENAPI_MISSING_PLUGIN' /* || error.code === 'UNKNOWN_XENAPI_PLUGIN_FUNCTION' */) {
```
Is this commented part intended to stay commented?

b-Nollet commented on packages/xo-server-sdn-controller/src/openflow-plugin.js:
```diff
@@ -0,0 +1,54 @@
+      const host = PIF.host
+      return host.$xapi.call('host.call_plugin', host.$ref, PLUGIN_NAME, method, { ...parameters, bridge })
+    })
+  }
+
+  async addRule({ vif, allow, protocol, ipRange, direction, port }) {
+    const network = vif.$network
+    const mac = vif.MAC
+    return this.#callPluginOnAllNetwork(network, 'add-flow', {
+      mac,
+      allow: allow ? 'true' : 'false',
+      protocol,
+      ipRange,
+      direction,
+      port,
+    })
+  }
```
I think is will look simpler like this:
```suggestion
  async addRule({ vif, allow, protocol, ipRange, direction, port }) {
    return this.#callPluginOnAllNetwork(vif.$network, 'add-flow', {
      mac: vif.MAC,
      allow: allow ? 'true' : 'false',
      protocol,
      ipRange,
      direction,
      port,
    })
  }
```

b-Nollet commented on packages/xo-server-sdn-controller/src/openflow-plugin.js:
```diff
@@ -0,0 +1,54 @@
+      direction,
+      port,
+    })
+  }
+
+  async deleteRule({ vif, allow, protocol, ipRange, direction, port }) {
+    const network = vif.$network
+    const mac = vif.MAC
+    return this.#callPluginOnAllNetwork(network, 'del-flow', {
+      mac,
+      allow: allow ? 'true' : 'false',
+      protocol,
+      ipRange,
+      direction,
+      port,
+    })
+  }
```
Same suggestion on `addRule`:
```suggestion
  async deleteRule({ vif, allow, protocol, ipRange, direction, port }) {
    return this.#callPluginOnAllNetwork(vif.$network, 'del-flow', {
      mac: vif.MAC,
      allow: allow ? 'true' : 'false',
      protocol,
      ipRange,
      direction,
      port,
    })
  }
```

b-Nollet commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -602,7 +628,39 @@ class SDNController extends EventEmitter {
+  async _addNetworkRule({ networkId, allow, direction, ipRange = '', port, protocol }) {
+    try {
+      const network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.PIFs.length > 0, 'Network needs to be plugged to delete  rule')
```
Typo: you probably meant "to delete a rule", or "to delete rules" (+ there is a double space)

b-Nollet commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -612,6 +638,38 @@ class SDNController extends EventEmitter {
+        newNetworkRules.push(stringRule)
+        await network.update_other_config('xo:sdn-controller:of-rules', JSON.stringify(newNetworkRules))
+      }
+    } catch (error) {
+      log.error('Error while adding Network OF rule', {
+        error,
+        networkId,
+        allow,
+        protocol,
+        port,
+        ipRange,
+        direction,
+      })
```
Is it intended that the error is ignored after being logged? (i.e. when calling this function, we have no way to tell if it ended successfully)
Same thing for `_deleteNetworkOfRule`

fbeauchamp replied:
that was the case in the original addRule implementation

b-Nollet commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -651,7 +709,52 @@ class SDNController extends EventEmitter {
+  async _deleteNetworkOfRule({ direction, ipRange = '', port, protocol, networkId }, updateOtherConfig = true) {
+    try {
+      let network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.PIFs.length > 0, 'Network needs to be plugged to delete  rule')
```
Same typo in `_addNetworkRule`

fbeauchamp committed:
fix following review

fbeauchamp committed:
fix: align with plugin api

b-Nollet commented on packages/xo-server-sdn-controller/src/openflow-plugin.js:
```diff
@@ -0,0 +1,58 @@
+          bridge: $network.bridge,
+        })
+        const json = JSON.parse(response)
+        strictEqual(json.returncode, 0, `plugin check should have a return code of 0 to succeed, got ${response}`)
```
Shouldn't it be this?
```suggestion
        strictEqual(json.returncode, 0, `plugin check should have a return code of 0 to succeed, got ${json.returncode}`)
```
or maybe 
```suggestion
        strictEqual(json.returncode, 0, `plugin check should have a return code of 0 to succeed, got ${json}`)
```

b-Nollet commented on packages/xo-server-sdn-controller/src/openflow-plugin.js:
```diff
@@ -0,0 +1,58 @@
+        const response = await host.$xapi.call('host.call_plugin', host.$ref, PLUGIN_NAME, 'dump-flows', {
+          bridge: $network.bridge,
+        })
+        const json = JSON.parse(response)
```
Are we sure this won't throw or that it's ok if it throws here?

fbeauchamp replied:
I think it's ok if it ails during the check that is used to check if the plugin is working as intended

fbeauchamp committed:
Update packages/xo-server-sdn-controller/src/openflow-plugin.js

Co-authored-by: Bastien Nollet <150058367+b-Nollet@users.noreply.github.com>

fbeauchamp committed:
feat(network): remove port for ICMP filters

fbeauchamp committed:
log

fbeauchamp committed:
fix host

b-Nollet reviewed (changes requested):
Is this normal that we still have several `console.log()`? I'm genuinely asking, as they were all added in the same commit (but the CI does not like it)

fbeauchamp committed:
changelog

fbeauchamp committed:
cleanup

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -1239,13 +1239,12 @@ class SDNController extends EventEmitter {
+      channel = this.ofChannels[host.$ref] = instantiateController(host)
     }
-
+    await channel
```
Is it intentional to store a promise? Even if you wait `channel`, a resolved promise will be stored, not just the result.
Additionally, in case your promise throw, `this.ofChannels[host.$ref]` is not cleaned up, so you won't try to instantiate another controller on the next call

Maybe something like:
```js
channel = await instantiateController(host)
this.ofChannels[host.$ref] = channel
```

```js
// E.g.
const foo = {};

async function bar() {
  await new Promise((resolve) => setTimeout(resolve, 3000));
  return "bar";
}

(async () => { 
  const p = foo["p"] = bar();
  await p
  console.log(foo["p"]); // Promise { 'bar' }
  console.log(p); // Promise { 'bar' }

  const d = await bar()
  foo["d"] = d;
  console.log(foo["d"]); // bar
  console.log(d); // bar
})()
```

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -612,6 +638,38 @@ class SDNController extends EventEmitter {
+    try {
+      const network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.PIFs.length > 0, 'Network needs to be plugged to delete  rule')
+      const channel = this._getOrCreateOfChannel(network.PIFs[0].host)
```
IMO, `network.PIFs` return an array of opaqueRef. Did you mean `network.$PIFs`?

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -660,6 +718,51 @@ class SDNController extends EventEmitter {
+      let network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.PIFs.length > 0, 'Network needs to be plugged to delete  rule')
+
+      const channel = await this._getOrCreateOfChannel(network.PIFs[0].host)
```
same

MathieuRA commented on packages/xo-server-sdn-controller/src/openflow-plugin.js:
```diff
@@ -0,0 +1,48 @@
+
+  async #callPluginOnAllNetwork(network, method, parameters) {
+    const bridge = this.#getBridge(network)
+    return asyncEach(network.PIFs, async PIF => {
```
same

b-Nollet reviewed (approved):
Nothing to add

fbeauchamp committed:
use  , not PIFs

fbeauchamp committed:
fix promises

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -602,7 +628,39 @@ class SDNController extends EventEmitter {
+  async _addNetworkRule({ networkId, allow, direction, ipRange = '', port, protocol }) {
+    try {
+      const network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.$PIFs.length > 0, 'Network needs to be plugged to delete a rule')
```
```suggestion
      assert(network.$PIFs.length > 0, 'Network needs to be plugged to add a rule')
```

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -602,7 +628,39 @@ class SDNController extends EventEmitter {
+  async _addNetworkRule({ networkId, allow, direction, ipRange = '', port, protocol }) {
+    try {
+      const network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.$PIFs.length > 0, 'Network needs to be plugged to delete a rule')
```
How about adding a test that ensures at least one PIF is attached?

fbeauchamp replied:
the rule has been added

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -602,7 +628,39 @@ class SDNController extends EventEmitter {
+      const network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.$PIFs.length > 0, 'Network needs to be plugged to delete a rule')
+      const channel = this._getOrCreateOfChannel(network.$PIFs[0].host)
+      await channel.addNetworkRule(network, allow, protocol, port, ipRange, direction)
```
```suggestion
      await channel.addNetworkRule({network, allow, protocol, port, ipRange, direction})
```

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -602,7 +628,39 @@ class SDNController extends EventEmitter {
+      await channel.addNetworkRule(network, allow, protocol, port, ipRange, direction)
+      const networkRules = network.other_config['xo:sdn-controller:of-rules']
+      const newNetworkRules = networkRules !== undefined ? JSON.parse(networkRules) : []
+      const stringRule = JSON.stringify({
+        allow,
+        protocol,
+        port,
+        ipRange,
+        direction,
+      })
```
Do you think it would be a good idea to inject an ID?
That way, if you want to delete a rule, simply use the identifier instead of all the parameters.

fbeauchamp replied:
I think this is a good idea, but this will lead to a big refactoring. We'll keep this idea when a decision is taken on this plugin future ( rework / rewrite/ )

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -602,7 +628,39 @@ class SDNController extends EventEmitter {
+      const network = this._xo.getXapiObject(this._xo.getObject(networkId, 'network'))
+      assert(network.$PIFs.length > 0, 'Network needs to be plugged to delete a rule')
+      const channel = this._getOrCreateOfChannel(network.$PIFs[0].host)
+      await channel.addNetworkRule(network, allow, protocol, port, ipRange, direction)
```
`addNetworkRule` do not exist on `openflow-channel`

fbeauchamp replied:
yes this is expected to fail without the plugin

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -651,7 +709,52 @@ class SDNController extends EventEmitter {
+      assert(network.$PIFs.length > 0, 'Network needs to be plugged to delete a rule')
+
+      const channel = await this._getOrCreateOfChannel(network.$PIFs[0].host)
+      await channel.deleteNetworkRule({ network, protocol, port, ipRange, direction })
```
`deleteNetworkRule` do not exist in `OpenflowChannel`

fbeauchamp replied:
yes this is expected to fail without the plugin

MathieuRA commented on packages/xo-server-sdn-controller/src/index.js:
```diff
@@ -1239,13 +1351,15 @@ class SDNController extends EventEmitter {
+      // this ensure only one channel is create in parallel
+      channel = this.ofChannels[host.$ref] = instantiateController(host)
+    }
+    if (this.ofChannels[host.$ref].then) {
```
Maybe use `isPromise` from `node:util/types`?

fbeauchamp replied:
is it more precise or handle more edge cases than checking for the .then property ?

fbeauchamp committed:
fix(sdn-controller): harden plugin detection

change algorithm from "all the network should be valid" to
"at least one should be valid " to test if the plugin is installed

fbeauchamp committed:
fix(sdn-controller): rename plugin

CI don't like dashes

fbeauchamp committed:
fix(sdn-controller): typo on log method

fbeauchamp committed:
fix(sdn-controller): force ports as string in case of xapi-plugin

Calling the plugin with integer raises an error, forcing port as string
but only if not undefined.

fbeauchamp committed:
fix(sdn-controller): properly find PIFs when create cross pool networks

PIF detection was fixed in 137837cdab310dc1c1495a9230d47e90d3a50ae2 but breaks the cross pool private
network cases, this fixes that properly using the center when not in the
current pool.

fbeauchamp committed:
feat(sdn-controller): use config to define if we use plugin or direct

fbeauchamp committed:
fix following mathieu's review

fbeauchamp committed:
fix: missing await

fbeauchamp committed:
more debug while missing PIFs

fbeauchamp committed:
use object instead of id

fbeauchamp committed:
allow must be a string for the api call

fbeauchamp committed:
better host loading for network rules

fbeauchamp committed:
iprange and protocol are not optionals

fbeauchamp committed:
fix optional parameters

fbeauchamp committed:
fix optional parameters

fbeauchamp committed:
better handling of allow property

MathieuRA reviewed (approved):
to tested

fbeauchamp committed:
changelog

fbeauchamp committed:
typo
